Insert Metrics Tracker in the toolchain

Tomcli opened the pull request:
* Add tracker to tracker metrics in the toolchain
* Add option to disable tracker
* Add environment variables to capture tracker option and chatbot name without timestamp.
* Add privacy notice

Tomcli committed:
Insert Tracker

* Add tracker to tracker metrics in the toolchain
* Add privacy notice

Tomcli commented:
I think my branch doesn't have the cf login credential to pass the build.

Tomcli committed:
Add deployment tracker badge

stevemar commented on .bluemix/pipeline.yml:
```diff
@@ -74,6 +80,31 @@ stages:
+      # Tracking Metrics
+      if [ "yes" != "$(echo "$DISABLE_METRIC" | awk '{print tolower($0)}')" ]; then
+      space_id=$(cf space ${CF_SPACE} --guid)
+      service_id=$(cf service Bot\ Asset\ Exchange\ Workspaces --guid)
```
nice, didn't know the `--guid` option exists! 

`cf service "Bot Asset Exchange Workspaces" --guid` also works if you want to avoid nasty backslahes

stevemar commented on README.md:
```diff
@@ -34,3 +36,17 @@ Underneath the covers, we hit the DevOps Service ``https://console.bluemix.net/d
+
+Sample DevOps pipeline that include this package may be configured to track deployments to [IBM Bluemix](https://www.bluemix.net/) and other platforms. The following information is sent to a [Deployment Tracker](https://github.com/IBM/metrics-collector-service) service on each deployment:
+
+* Space ID (`space_id`)
```
remove "(`space_id`)" here, the reader won't know what that means

stevemar commented on README.md:
```diff
@@ -34,3 +36,17 @@ Underneath the covers, we hit the DevOps Service ``https://console.bluemix.net/d
+
+## Privacy Notice
+
+Sample DevOps pipeline that include this package may be configured to track deployments to [IBM Bluemix](https://www.bluemix.net/) and other platforms. The following information is sent to a [Deployment Tracker](https://github.com/IBM/metrics-collector-service) service on each deployment:
```
I'd rephrase this as:

"The DevOps pipeline that is included in this package is configured by default with a [Deployment Tracker](https://github.com/IBM/metrics-collector-service) tool. The following information is sent when a user deploys an asset from the Bot Asset Exchange:"

stevemar reviewed (changes requested):
a few minor comments, but i need some time to actually try this out, too.

Tomcli committed:
update README and pipeline as requested

stevemar reviewed (commented):
sorry i haven't had the time to try this out yet @Tomcli

stevemar commented:
@Tomcli how did you test this out? Specifically, the UI changes? Was it adapted from somewhere else? Ideally I want to try this out but I'm not sure if we made the flow smart enough to use branches

Tomcli commented:
All the UI changes are in the deploy.json file. I added a new section that allows users to disable the tracker. You can create a new branch with my changes and try it out with any chatbot by adding `&branch=yourBranchName` at the end of the toolchain link. This way you should able to test the new UI and pipeline code in that branch.

stevemar commented on .bluemix/deploy.json:
```diff
@@ -32,6 +32,12 @@
+            "description": {
+                "$ref": "#/messages/deploy.metricsDescription"
+            },
+            "type": "string"
```
it would be great if this were a dropdown menu like the screenshot below. I looked at the docs and couldn't find the right option though -- the screenshot was from the watson conversation pipeline config

<img width="1058" alt="screen shot 2017-11-07 at 12 00 34 am" src="https://user-images.githubusercontent.com/4118756/32478059-6a578264-c350-11e7-86b7-5dcd1ca5baab.png">

stevemar replied:
maybe type: select? i'm not sure

stevemar commented on .bluemix/pipeline.yml:
```diff
@@ -74,6 +80,31 @@ stages:
       fi
+      
+      # Tracking Metrics
+      if [ "yes" != "$(echo "$DISABLE_METRIC" | awk '{print tolower($0)}')" ]; then
```
so as it stands now it's enabled by default

stevemar replied:
we should echo out some stuff here...

         echo "============================================================="
         echo "Adding deployment data to metrics tracker... "
         echo "============================================================="

and then again at the end if it was successful uploaded. I say this because I'm not seeing anything in the logs past the following:

```
==================================================
Created new workspace: WORKSPACE_ID: 645bdf74-c71a-4693-86ab-a860bb3eed1e
==================================================
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   310    0   310    0     0   2832      0 --:--:-- --:--:-- --:--:--  2844
{
  "name": "watson-banking-chatbot",
  "created": "2017-11-07T05:19:42.239Z",
  "updated": "2017-11-07T05:19:42.239Z",
  "language": "en",
  "metadata": {
    "runtime_version": "2016-09-20"
  },
  "description": "A cognitive banking chatbot",
  "workspace_id": "645bdf74-c71a-4693-86ab-a860bb3eed1e",
  "learning_opt_out": false,
  "status": "Training"
}

Finished: SUCCESS
```

stevemar commented on .bluemix/nls/messages.yml:
```diff
@@ -7,6 +7,7 @@ deploy.description: "sample toolchain"
 deploy.longDescription: "The Delivery Pipeline automates continuous deployment."
 deploy.appDescription: "The name of your Watson app"
 deploy.appName: "App name"
+deploy.metricsDescription : "The metrics tracker is for IBM to track metrics around deployments of sample applications to IBM Bluemix to measure the usefulness of our examples, so that we can continuously improve the content we offer to you. If you want to disable the metrics tracker, please put down YES."
```
We should state clearly the default behavior. Please change this to:

"The metrics tracker, which is enabled by default, is for IBM to track the number of times our examples are deployed to IBM Cloud. It is strictly used to measure the usefulness of our examples so that we can continuously improve the content we offer to you. If you want to disable the metrics tracker, please put down YES."

stevemar commented on README.md:
```diff
@@ -1,3 +1,5 @@
+![Bluemix Deployments](https://metrics-tracker.mybluemix.net/stats/3dafda64b9efbd06391d5f090bd89113/badge.svg)
```
i don't think we need this, but meh.

stevemar commented on README.md:
```diff
@@ -34,3 +36,17 @@ Underneath the covers, we hit the DevOps Service ``https://console.bluemix.net/d
+* Conversation Service GUID
+* Chatbot Name
+
+This data is collected from the Bluemix DevOps pipeline environment variables in IBM Bluemix. This data is used by IBM to track metrics around deployments of sample applications to IBM Bluemix to measure the usefulness of our examples, so that we can continuously improve the content we offer to you. Only deployments of sample applications that include code to ping the Deployment Tracker service w ...
```
replace Bluemix with IBM Cloud here, mention that it's enabled by default

stevemar commented on README.md:
```diff
@@ -34,3 +36,17 @@ Underneath the covers, we hit the DevOps Service ``https://console.bluemix.net/d
+
+## Disabling Deployment Tracking
+
+To disable the tracker, Simply put down `yes` in *Disable Metrics Tracker* under the **Delivery Pipeline** section when you create your Bluemix DevOps pipeline.
```
i would say "opt out" instead of "disable" here

stevemar reviewed (commented):
Thanks @Tomcli -- i left a few more suggestions around wording and logging that I'd like cleaned up before this merges.

I also have a UI suggestion, see if you can get a dropdown menu working instead of a text field for the metrics tracker.

I'll go over this with anamita and show her the UI changes.

I tested this by going to: https://console.bluemix.net/devops/setup/deploy?repository=https%3A//github.com/Tomcli/chatbot-deployer&chatbotName=Simple%20Banking%20Chatbot&chatbotWorkspaceURL=http://developer.ibm.com/code/wp-content/uploads/sites/118/2017/09/banking.json&cm_sp=dw-bluemix-_-code-_-devcenter

<img width="1037" alt="screen shot 2017-11-06 at 11 58 52 pm" src="https://user-images.githubusercontent.com/4118756/32478601-9e616a0e-c353-11e7-9104-1b1dab1b659f.png">

Tomcli committed:
Update tracker with a dropdown selection. (#2)

* Update tracker with a dropdown selection
* Update privacy notice to IBM Cloud

Tomcli commented:
Thank you for your feedback @stevemart. I updated the toolchain UI to a dropdown and added some logs for the tracker. Please use the below [link](https://console.bluemix.net/devops/setup/deploy?repository=https%3A//github.com/Tomcli/chatbot-deployer&chatbotName=Simple%20Banking%20Chatbot&chatbotWorkspaceURL=http://developer.ibm.com/code/wp-content/uploads/sites/118/2017/09/banking.json&cm_sp=dw-bluemix-_-code-_-devcenter&branch=tracker) to test the new UI and pipeline code because I am merging from my `tracker` branch.
- https://console.bluemix.net/devops/setup/deploy?repository=https%3A//github.com/Tomcli/chatbot-deployer&chatbotName=Simple%20Banking%20Chatbot&chatbotWorkspaceURL=http://developer.ibm.com/code/wp-content/uploads/sites/118/2017/09/banking.json&cm_sp=dw-bluemix-_-code-_-devcenter&branch=tracker

stevemar commented on .bluemix/pipeline.yml:
```diff
@@ -74,6 +80,40 @@ stages:
+      EOF
+      )
+
+      callback=$(curl -d "$data" -H "Content-Type: application/json" -X POST http://metrics-tracker.mybluemix.net/api/v1/track)
```
one last comment, but i think the URL should be a defined variable

stevemar reviewed (approved):
one super final comment, just flip the wording so that if the user selects "YES" they are enabling the tracker, and if they select "NO" they are disabling it. 

positive answers, like "YES", should do positive things.

Tomcli committed:
Set positive option to YES, set tracker endpoint as env var (#3)

stevemar commented on .bluemix/pipeline.yml:
```diff
@@ -105,7 +108,7 @@ stages:
 
-      callback=$(curl -d "$data" -H "Content-Type: application/json" -X POST http://metrics-tracker.mybluemix.net/api/v1/track)
+      callback=$(curl -d "$data" -H "Content-Type: application/json" -X POST $TRACKER_URL)
       if [[ "$callback" == *"true"* ]]; then 
```
you probably want to indent this if block

stevemar commented on .bluemix/pipeline.yml:
```diff
@@ -82,7 +85,7 @@ stages:
       # Tracking Metrics
-      if [ "YES" != "$DISABLE_METRIC" ]; then
+      if [ "NO" != "$DISABLE_METRIC" ]; then
       echo "============================================================="
```
i just realized this block is not indented.

Tomcli replied:
Sorry I didn't indent this block because the [EOF](https://github.com/Tomcli/chatbot-deployer/blob/tracker/.bluemix/pipeline.yml#L108) portion of the code doesn't work if there are spaces after it. I will try to find if there's another solution.

stevemar reviewed (approved):
@Tomcli merging this change, try to clean up the indentation around the if blocks, they're a bit out of whack
